README.md
# postcodes_io_api
A library that provides a Python interface to the postcodes_io API.

# Project URLs
 * Project Home : https://postcodes.io 
 * Project Doc  : https://postcodes.io/docs

# Example
```
  import postcodes_io_api
  api  = postcodes_io_api.Api(debug_http=True)
  data = api.get_postcode('SW112EF')
```

# Api
```python
Api(self, debug_http=False, timeout=None, base_url=None)
```

## is_postcode_valid
```python
Api.is_postcode_valid(self, postcode)
```

This method validates post_code
* **:param postcode** - postcode to check i.e. 'SW112EF'
* **:return** - True if postcode is valid False if postcode is invalid

```
  is_valid = api.is_postcode_valid('SW112EF')

```

## is_postcode_terminated
```python
Api.is_postcode_terminated(self, postcode)
```

* **:param postcode** - postcode to check i.e. 'SW112ZW'
* **:return** True if postcode is terminated or False otherwise

```
  is_terminated = api.is_postcode_terminated('SW112EF')

```

## get_postcode
```python
Api.get_postcode(self, postcode)
```

This method returns data for post_code
* **:param postcode** - postcode to check i.e. 'SW112EF'
* **::return** - postcode detailed data
```
  data = api.get_postcode('SW112EF')

```

## get_nearest_postcodes_for_postcode
```python
Api.get_nearest_postcodes_for_postcode(self, **kwargs)
```

* **kwargs**
* **:param postcode** - postcode
* **:param limit** - (not required) Limits number of postcodes matches to return. Defaults to 10. Needs to be less than 100.
* **:param radius** -  (not required) Limits number of postcodes matches to return. Defaults to 100m. Needs to be less than 2,000m.
* **:return:** - list of nearest postcodes data

```
  data = api.get_nearest_postcodes_for_postcode(postcode='SW112EF',limit=2)

```

## get_nearest_postcodes_for_coordinates
```python
Api.get_nearest_postcodes_for_coordinates(self, **kwargs)
```

* **kwargs**
* **:param latitude** - (required) Latitude
* **:param longitude** - (required) Longitude
* **:param limit** - (not required) Limits number of postcodes matches to return. Defaults to 10. Needs to be less than 100.
* **:param radius** -  (not required) Limits number of postcodes matches to return. Defaults to 100m. Needs to be less
* **:return:** - list of nearest postcodes data

```
  data = api.get_nearest_postcodes_for_coordinates(latitude=51.466324,longitude=-0.173606,limit=2)

```


## get_bulk_postcodes
```python
Api.get_bulk_postcodes(self, postcodes_list)
```

* **:param postcodes_list** - list containing postcodes
* **:return** - list of postcode data

```
  postcode_list = ["SW112EF","HA97QP"]
  data = api.get_bulk_postcodes(postcode_list)

```

## get_bulk_reverse_geocode
```python
Api.get_bulk_reverse_geocode(self, payload_data)
```

* **:param payload_data** - dict with cordinates e.g.
```
payload_data = {
"geolocations":
[
    {
        "longitude": 0.629834723775309,
        "latitude": 51.7923246977375
    },
    {
        "longitude": -2.49690382054704,
        "latitude": 53.5351312861402,
        "radius": 1000,
        "limit": 5
    }
]
}
```
* **:return** - list of postcode data

```
    data = api.get_bulk_reverse_geocode(payload_data)
```


## get_random_postcode
```python
Api.get_random_postcode(self)
```

* **:return** - random postcode

```
    data = api.get_random_postcode()
```

## get_autocomplete_postcode
```python
Api.get_autocomplete_postcode(self, **kwargs)
```

* **kwargs**
* **:param postcode** - partial postcode
* **:param limit** - (not required) Limits number of postcodes matches to return. Defaults to 10. Needs to be less than 100.
* **:return** -  list of possible postcodes

```
    data = api.get_autocomplete_postcode(postcode='SW18',limit=2)
```

## get_outcode
```python
Api.get_outcode(self, outcode)
```

This method returns data for post_code
* **:param outcode** - postcode outward code to check i.e. 'KT1'
* **:return** - postcode detailed data

```
    data = api.get_outcode('KT1')
```

## get_nearest_outcodes_for_outcode
```python
Api.get_nearest_outcodes_for_outcode(self, **kwargs)
```

* **kwargs**
* **:param outcode** - outward code
* **:param limit** - (not required) Limits number of postcodes matches to return. Defaults to 10. Needs to be less than 100.
* **:param radius** - (not required) Limits number of postcodes matches to return. Defaults to 100m. Needs to be less than 2,000m.
* **:return:** - list of nearest postcodes data

```
    data = api.get_nearest_outcodes_for_outcode(outcode='KT1',limit=2)
```

## get_nearest_outcodes_for_coordinates
```python
Api.get_nearest_outcodes_for_coordinates(self, **kwargs)
```

* **kwargs**
* **:param latitude** - (required) Latitude
* **:param longitude** - (required) Longitude
* **:param limit** - (not required) Limits number of postcodes matches to return. Defaults to 10. Needs to be less than 100.
* **:param radius** -  (not required) Limits number of postcodes matches to return. Defaults to 100m. Needs to be less
* **:return:** - list of nearest outcodes data

```
    data = api.get_nearest_outcodes_for_coordinates(latitude=51.466324,longitude=-0.173606,limit=2)
```

# Changelog - postcodes_io_api
A library that provides a Python interface to the postcodes_io API.

# 0.0.4 (12-01-2019)
 * fixed packaging issues for python 2

# 0.0.3 (12-01-2019)
 * added suport for python 2

# 0.0.2 (07-01-2019)
 * fixed requirements and setup file

# 0.0.1 (07-01-2019)
 * initial release



CHANGES.md
# Changelog - postcodes_io_api
A library that provides a Python interface to the postcodes_io API.

# 0.0.3 (12-01-2019)
 * added suport for python 2

# 0.0.2 (07-01-2019)
 * fixed requirements and setup file

# 0.0.1 (07-01-2019)
 * initial release


tests/test.py
import unittest

import postcodes_io_api
import random


class PostcodeIOTest(unittest.TestCase):
    def setUp(self):
        self.api_client = postcodes_io_api.Api()
        self.latitude = 51.466324
        self.longitude = -0.173606
        self.VALID_OUTCODES = ['KT19', 'HA9', 'SW19', 'IV2', 'PH16']
        self.VALID_POSTCODES = ['L40TH', 'SW151JF', 'SW195AG', 'KT185LQ', 'NW87JY', 'RG11LZ']
        self.INVALID_POSTCODES = ['HA9997QP', 'SW19OSZ', 'HA899NX', 'RG101LZ', 'SW12EF', 'L500QD']
        self.TERMINATED_POSTCODES = ['SW112ZW', 'HA89NX', 'IV23EH', 'BA12QT', 'OX13LA', 'SE109DB']
        self.reverse_geocode_data = {
            "geolocations": [{
                "longitude": 0.629834723775309,
                "latitude": 51.7923246977375
            }, {
                "longitude": -2.49690382054704,
                "latitude": 53.5351312861402,
                "radius": 1000,
                "limit": 5
            }]
        }

    def test_is_postcode_valid(self):
        self.assertTrue(self.api_client.is_postcode_valid(random.choice(self.VALID_POSTCODES)))
        self.assertFalse(self.api_client.is_postcode_valid(random.choice(self.INVALID_POSTCODES)))
        self.assertFalse(self.api_client.is_postcode_valid(random.choice(self.TERMINATED_POSTCODES)))

    def test_is_postcode_terminated(self):
        self.assertFalse(self.api_client.is_postcode_terminated(random.choice(self.VALID_POSTCODES)))
        self.assertFalse(self.api_client.is_postcode_terminated(random.choice(self.INVALID_POSTCODES)))
        self.assertTrue(self.api_client.is_postcode_terminated(random.choice(self.TERMINATED_POSTCODES)))

    def test_get_postcode(self):
        valid_data = self.api_client.get_postcode(random.choice(self.VALID_POSTCODES))
        self.assertEqual(200, valid_data.get('status', None), "Checking for valid postcode")
        invalid_data = self.api_client.get_postcode(random.choice(self.INVALID_POSTCODES))
        self.assertEqual(404, invalid_data.get('status', None), "Checking for invalid postcode")
        terminated_data = self.api_client.get_postcode(random.choice(self.TERMINATED_POSTCODES))
        self.assertEqual(404, terminated_data.get('status', None), "Checking for terminated postcode")

    def test_get_nearest_postcodes_for_postcode(self):
        data = self.api_client.get_nearest_postcodes_for_postcode(postcode=random.choice(self.VALID_POSTCODES), limit=2)
        self.assertEqual(200, data.get('status', None), "Checking for valid postcode")

    def test_get_nearest_postcodes_for_coordinates(self):
        data = self.api_client.get_nearest_postcodes_for_coordinates(latitude=self.latitude, longitude=self.longitude,
                                                                     limit=3)
        self.assertEqual(200, data.get('status', None))

    def test_get_bulk_postcodes(self):
        bulk_postcodes = [random.choice(self.VALID_POSTCODES), random.choice(self.VALID_POSTCODES)]
        data = self.api_client.get_bulk_postcodes(bulk_postcodes)
        self.assertEqual(200, data.get('status', None), "Checking for valid postcode")
        self.assertEqual(2, len(data.get('result', [])), "Checking for result count")

    def test_get_bulk_reverse_geocode(self):
        data = self.api_client.get_bulk_reverse_geocode(self.reverse_geocode_data)
        self.assertEqual(200, data.get('status', None))
        self.assertEqual(2, len(data.get('result', [])), "Checking for result count")

    def test_get_random_postcode(self):
        data = self.api_client.get_random_postcode()
        self.assertEqual(200, data.get('status', None))

    def test_get_autocomplete_postcode(self):
        data = self.api_client.get_autocomplete_postcode(postcode=random.choice(self.VALID_OUTCODES), limit=3)
        self.assertEqual(200, data.get('status', None))
        self.assertEqual(3, len(data.get('result', [])), "Checking for result count")

    def test_get_outcode(self):
        data = self.api_client.get_outcode(random.choice(self.VALID_OUTCODES))
        self.assertEqual(200, data.get('status', None), "Checking for valid outcode")

    def test_get_nearest_outcodes_for_outcode(self):
        data = self.api_client.get_nearest_outcodes_for_outcode(outcode=random.choice(self.VALID_OUTCODES))
        self.assertEqual(200, data.get('status', None), "Checking for valid outcode")

    def test_get_nearest_outcodes_for_coordinates(self):
        data = self.api_client.get_nearest_outcodes_for_coordinates(latitude=self.latitude, longitude=self.longitude,
                                                                    limit=3)
        self.assertEqual(200, data.get('status', None))
        self.assertEqual(3, len(data.get('result', [])), "Checking for result count")

    def tearDown(self):
        """
            : close api_client session to avoid 'ResourceWarning: unclosed <socket.socket'
        """
        self.api_client._session.close()


example.py
#!/usr/bin/env python

import postcodes_io_api

api = postcodes_io_api.Api(debug_http=True)

postcode1 = 'SW112EF'
postcode2 = 'SW112222'


data1 = api.get_postcode(postcode1)
print("\n\n",data1)
#data2 = api.get_postcode(postcode2)
#print(data2)

#run only certain tests
#python setup.py test -s tests.test.PostcodeIOTest.test_get_postcode

"""
lat = 51.466324
lon = -0.173606
data = api.get_nearest_postcodes_for_coordinates(latitude=lat,longitude=lon,limit=2)
print(data)
"""
"""
#data = api.get_bulk_postcodes({"postcodes":["OX49 5NU","M32 0JG","NE30 1DP"]})
list = ["SW112EF"]
data = api.get_bulk_postcodes(list)
print("\n")
print(data)
"""

"""
if api.validate_postcode(postcode1):
    print("postcode ="+ postcode1+ " is valid")
else:
    print("postcode =" + postcode1 + " is invalid")

if api.validate_postcode(postcode2):
    print("postcode ="+ postcode2+ " is valid")
else:
    print("postcode =" + postcode2 + " is invalid")
"""

"""
#get random postcode
postcode_data = api.get_random_postcode()
print("random_postcode\n\n",postcode_data)
"""
"""
#get get_nearest_postcodes_for_postcode
data = api.get_nearest_postcodes_for_postcode(postcode='SW112EF')
print("\n",data)
"""

"""
#get
data =api.get_autocomplete_postcode(postcode='SW112E')
print("\n\n",data)
"""
"""
terminated_postcode = 'SW11 2ZW'
#terminated_postcode = 'SW112EF'
if api.is_postcode_terminated(terminated_postcode):
    print(terminated_postcode+" is_terminated")
else:
    print(terminated_postcode+" is_working_postcode")
"""

"""
payload_data = {
    "geolocations": [{
        "longitude": 0.629834723775309,
        "latitude": 51.7923246977375
    }, {
        "longitude": -2.49690382054704,
        "latitude": 53.5351312861402,
        "radius": 1000,
        "limit": 5
    }]
}
data = api.get_bulk_reverse_geocode(payload_data)
print("\n\n",data)
"""

"""
outcode = 'KT1'
data  = api.get_outcode(outcode)
print("\n\n",data)
"""

"""
outcode = 'KT1'
data = api.get_nearest_outcodes_for_outcode(outcode=outcode,limit=2)
print("\n\n",data)
"""

"""
lat = 51.466324
lon = -0.173606
data = api.get_nearest_outcodes_for_coordinates(latitude=lat,longitude=lon,limit=1)
print("\n\n",data)
"""



postcodes_io_api/api.py
import requests
import logging
import json

try:
    # python 3
    import http.client as http_client
    from urllib.parse import urlparse, urlunparse, urlencode
except ImportError:
    # python 2
    import httplib as http_client
    from urlparse import urlparse, urlunparse
    from urllib import urlencode

API_BASE_URL = 'http://api.postcodes.io'


class Api(object):
    def __init__(self, debug_http=False, timeout=None, base_url=None):
        """
        Instantiate a new postcodes_io.Api object

        * **:param debug_http**  - boolean (optional) To enable logging for http requests
        * **:param timeout **  - int Set timeout for http/https requests

        """
        self._debug_http = debug_http
        self._timeout = timeout
        if base_url:
            self.base_url = base_url
        else:
            self.base_url = API_BASE_URL

        if debug_http:
            http_client.HTTPConnection.debuglevel = 1
            logging.basicConfig()  # initialize logging
            logging.getLogger().setLevel(logging.DEBUG)
            requests_log = logging.getLogger("requests.packages.urllib3")
            requests_log.setLevel(logging.DEBUG)
            requests_log.propagate = True
        self._session = requests.Session()

    def is_postcode_valid(self, postcode):
        """
        This method validates post_code
        * **:param postcode** - postcode to check i.e. 'SW112EF'
        * **:return** - True if postcode is valid False if postcode is invalid

        ```
          is_valid = api.is_postcode_valid('SW112EF')

        ```
        """
        url = '/postcodes/{postcode}/validate'.format(postcode=postcode)
        response = self._make_request('GET', url)
        data = self._parse_json_data(response.content.decode('utf-8'))
        return True if response.status_code == 200 and data.get('result') else False

    def is_postcode_terminated(self, postcode):
        """
        * **:param postcode** - postcode to check i.e. 'SW112ZW'
        * **:return** True if postcode is terminated or False otherwise

        ```
          is_terminated = api.is_postcode_terminated('SW112EF')

        ```
        """
        url = '/terminated_postcodes/{postcode}'.format(postcode=postcode)
        response = self._make_request('GET', url)
        data = self._parse_json_data(response.content.decode('utf-8'))
        return True if response.status_code == 200 and data.get('result', {}).get('postcode') else False

    def get_postcode(self, postcode):
        """
        This method returns data for post_code
        * **:param postcode** - postcode to check i.e. 'SW112EF'
        * **::return** - postcode detailed data
        ```
          data = api.get_postcode('SW112EF')

        ```
        """
        url = '/postcodes/{postcode}'.format(postcode=postcode)
        response = self._make_request('GET', url)
        data = self._parse_json_data(response.content.decode('utf-8'))
        return data

    def get_nearest_postcodes_for_postcode(self, **kwargs):
        """
        * **kwargs**
        * **:param postcode** - postcode
        * **:param limit** - (not required) Limits number of postcodes matches to return. Defaults to 10. Needs to be less than 100.
        * **:param radius** -  (not required) Limits number of postcodes matches to return. Defaults to 100m. Needs to be less than 2,000m.
        * **:return:** - list of nearest postcodes data

        ```
          data = api.get_nearest_postcodes_for_postcode(postcode='SW112EF',limit=2)

        ```
        """
        url = '/postcodes/{postcode}/nearest'.format(postcode=kwargs.pop('postcode', ''))
        response = self._make_request('GET', url, data=kwargs)
        data = self._parse_json_data(response.content.decode('utf-8'))
        return data

    def get_nearest_postcodes_for_coordinates(self, **kwargs):
        """
        * **kwargs**
        * **:param latitude** - (required) Latitude
        * **:param longitude** - (required) Longitude
        * **:param limit** - (not required) Limits number of postcodes matches to return. Defaults to 10. Needs to be less than 100.
        * **:param radius** -  (not required) Limits number of postcodes matches to return. Defaults to 100m. Needs to be less
        * **:return:** - list of nearest postcodes data

        ```
          data = api.get_nearest_postcodes_for_coordinates(latitude=51.466324,longitude=-0.173606,limit=2)

        ```

        """
        if kwargs.get('latitude') and kwargs.get('longitude'):
            url = '/postcodes'
            response = self._make_request('GET', url, data=kwargs)
            data = self._parse_json_data(response.content.decode('utf-8'))
            return data

    def get_bulk_postcodes(self, postcodes_list):
        """
        * **:param postcodes_list** - list containing postcodes
        * **:return** - list of postcode data

        ```
          postcode_list = ["SW112EF","HA97QP"]
          data = api.get_bulk_postcodes(postcode_list)

        ```
        """
        url = '/postcodes'
        response = self._make_request('POST', url, json={'postcodes': postcodes_list})
        data = self._parse_json_data(response.content.decode('utf-8'))
        return data

    def get_bulk_reverse_geocode(self, payload_data):
        """
        * **:param payload_data** - dict with cordinates e.g.
        ```
        payload_data = {
        "geolocations":
        [
            {
                "longitude": 0.629834723775309,
                "latitude": 51.7923246977375
            },
            {
                "longitude": -2.49690382054704,
                "latitude": 53.5351312861402,
                "radius": 1000,
                "limit": 5
            }
        ]
        }
        ```
        * **:return** - list of postcode data

        ```
            data = api.get_bulk_reverse_geocode(payload_data)
        ```

        """
        url = '/postcodes'
        response = self._make_request('POST', url, json=payload_data)
        data = self._parse_json_data(response.content.decode('utf-8'))
        return data

    def get_random_postcode(self):
        """
        * **:return** - random postcode

        ```
            data = api.get_random_postcode()
        ```
        """
        url = '/random/postcodes'
        response = self._make_request('GET', url)
        data = self._parse_json_data(response.content.decode('utf-8'))
        return data

    def get_autocomplete_postcode(self, **kwargs):
        """
        * **kwargs**
        * **:param postcode** - partial postcode
        * **:param limit** - (not required) Limits number of postcodes matches to return. Defaults to 10. Needs to be less than 100.
        * **:return** -  list of possible postcodes

        ```
            data = api.get_autocomplete_postcode(postcode='SW18',limit=2)
        ```
        """
        url = '/postcodes/{postcode}/autocomplete'.format(postcode=kwargs.pop('postcode', ''))
        response = self._make_request('GET', url, data=kwargs)
        data = self._parse_json_data(response.content.decode('utf-8'))
        return data

    def get_outcode(self, outcode):
        """
        This method returns data for post_code
        * **:param outcode** - postcode outward code to check i.e. 'KT1'
        * **:return** - postcode detailed data

        ```
            data = api.get_outcode('KT1')
        ```
        """
        url = '/outcodes/{outcode}'.format(outcode=outcode)
        response = self._make_request('GET', url)
        data = self._parse_json_data(response.content.decode('utf-8'))
        return data

    def get_nearest_outcodes_for_outcode(self, **kwargs):
        """
        * **kwargs**
        * **:param outcode** - outward code
        * **:param limit** - (not required) Limits number of postcodes matches to return. Defaults to 10. Needs to be less than 100.
        * **:param radius** - (not required) Limits number of postcodes matches to return. Defaults to 100m. Needs to be less than 2,000m.
        * **:return:** - list of nearest postcodes data

        ```
            data = api.get_nearest_outcodes_for_outcode(outcode='KT1',limit=2)
        ```
        """
        url = '/outcodes/{outcode}/nearest'.format(outcode=kwargs.pop('outcode', ''))
        response = self._make_request('GET', url, data=kwargs)
        data = self._parse_json_data(response.content.decode('utf-8'))
        return data

    def get_nearest_outcodes_for_coordinates(self, **kwargs):
        """
        * **kwargs**
        * **:param latitude** - (required) Latitude
        * **:param longitude** - (required) Longitude
        * **:param limit** - (not required) Limits number of postcodes matches to return. Defaults to 10. Needs to be less than 100.
        * **:param radius** -  (not required) Limits number of postcodes matches to return. Defaults to 100m. Needs to be less
        * **:return:** - list of nearest outcodes data

        ```
            data = api.get_nearest_outcodes_for_coordinates(latitude=51.466324,longitude=-0.173606,limit=2)
        ```
        """
        if kwargs.get('latitude') and kwargs.get('longitude'):
            url = '/outcodes'
            response = self._make_request('GET', url, data=kwargs)
            data = self._parse_json_data(response.content.decode('utf-8'))
            return data

    # TODO: add extra method to get haversine distance between postcodes using their coordinates
    def get_distance_between_postcodes(self, **kwargs):
        pass

    def _make_request(self, http_method, url, data=None, json=None):
        """
        :param http_method: http method i.e. GET, POST, PUT etc
        :param url: api endpoint url
        :param data: dictionary of key/value params
        :param json: json data
        :return:
        """
        if not data:
            data = {}

        response = 0
        if http_method == 'GET':
            url = self._build_url(url, extra_params=data)
            response = self._session.get(url, timeout=self._timeout)
        elif http_method == 'POST':
            url = self._build_url(url)
            if data:
                response = self._session.post(url, data=data, timeout=self._timeout)
            elif json:
                response = self._session.post(url, json=json, timeout=self._timeout)
        return response

    def _build_url(self, url, path_elements=None, extra_params=None):
        url = self.base_url + url
        (scheme, netloc, path, params, query, fragment) = urlparse(url)

        # Add extra_parameters to the query
        if extra_params and len(extra_params) > 0:
            extra_query = self._encode_parameters(extra_params)
            if query:
                query += '&' + extra_query
            else:
                query = extra_query
        return urlunparse((scheme, netloc, path, params, query, fragment))

    @staticmethod
    def _encode_parameters(parameters):
        """
        Return url encoded string in 'key=value&key=value' format
        """
        if not parameters or not isinstance(parameters, dict):
            return None
        return urlencode(parameters)

    @staticmethod
    def _parse_json_data(json_data):
        try:
            data = json.loads(json_data)
        except ValueError:
            data = json.loads('{"Error": "Unknown error while parsing response"}')
        return data
